indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialGenericSegmentMergeTask.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.druid.java.util.http.client.HttpClient;
+import org.apache.druid.timeline.partition.ShardSpec;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * {@link ParallelIndexTaskRunner} for the phase to merge generic partitioned segments in multi-phase parallel indexing.
+ */
+public class PartialGenericSegmentMergeTask extends PartialSegmentMergeTask<ShardSpec, GenericPartitionLocation>
+{
+  public static final String TYPE = "partial_index_generic_merge";
+
+  private final PartialGenericSegmentMergeIngestionSpec ingestionSchema;
+  private final Table<Interval, Integer, ShardSpec> createIntervalAndIntegerToShardSpec;
+
+  @JsonCreator
+  public PartialGenericSegmentMergeTask(
+      // id shouldn't be null except when this task is created by ParallelIndexSupervisorTask
+      @JsonProperty("id") @Nullable String id,
+      @JsonProperty("groupId") final String groupId,
+      @JsonProperty("resource") final TaskResource taskResource,
+      @JsonProperty("supervisorTaskId") final String supervisorTaskId,
+      @JsonProperty("numAttempts") final int numAttempts, // zero-based counting
+      @JsonProperty("spec") final PartialGenericSegmentMergeIngestionSpec ingestionSchema,
+      @JsonProperty("context") final Map<String, Object> context,
+      @JacksonInject IndexingServiceClient indexingServiceClient,
+      @JacksonInject IndexTaskClientFactory<ParallelIndexSupervisorTaskClient> taskClientFactory,
+      @JacksonInject @EscalatedClient HttpClient shuffleClient
+  )
+  {
+    super(
+        getOrMakeId(id, TYPE, ingestionSchema.getDataSchema().getDataSource()),
+        groupId,
+        taskResource,
+        supervisorTaskId,
+        ingestionSchema.getDataSchema(),
+        ingestionSchema.getIOConfig(),
+        ingestionSchema.getTuningConfig(),
+        numAttempts,
+        context,
+        indexingServiceClient,
+        taskClientFactory,
+        shuffleClient
+    );
+
+    this.ingestionSchema = ingestionSchema;
+    this.createIntervalAndIntegerToShardSpec = createIntervalAndIntegerToShardSpec(
+        ingestionSchema.getIOConfig().getPartitionLocations()
+    );
+  }
+
+  private static Table<Interval, Integer, ShardSpec> createIntervalAndIntegerToShardSpec(
+      List<GenericPartitionLocation> partitionLocations
+  )
+  {
+    Table<Interval, Integer, ShardSpec> intervalAndIntegerToShardSpec = HashBasedTable.create();
+
+    partitionLocations.forEach(
+        p -> intervalAndIntegerToShardSpec.put(p.getInterval(), p.getPartitionId(), p.getShardSpec())
+    );
+
+    return intervalAndIntegerToShardSpec;
+  }
+
+  @JsonProperty("spec")
+  private PartialGenericSegmentMergeIngestionSpec getIngestionSchema()
+  {
+    return ingestionSchema;
+  }
+
+  @Override
+  public String getType()
+  {
+    return TYPE;
+  }
+
+  @Override
+  ShardSpec createShardSpec(TaskToolbox toolbox, Interval interval, int partitionNum)
```

Comment:
Suggest to use `partitionId` instead of `partitionNum` because I think it's more clear. Same for other places.